src/test/scala/com/box/castle/router/kafkadispatcher/cache/CacheTest.scala

```diff
@@ -394,4 +394,38 @@ class CacheTest extends Specification with Mockito with MockBatchTools {
 
     }
   }
+
+  "Cache.getAll" should {
+    "return None if the cache is empty" in {
+      val cache = Cache(100)
+      // Check it returns none
+      cache.getAll(10, 100) shouldEqual None
+    }
+
+    "return None in case of cache miss" in {
+      val cache = Cache(300).add(createBatch(20, 1, 50))
+      // Check it returns none
+      cache.getAll(50, 100) shouldEqual None
+    }
+
+    "return all available contiguous messages in the cache starting with given offset that fit in buffersize" in {
+      var cache = Cache(3000)
+
+      val b1 = createBatch(20, 2, 65)
+      val b2 = createBatch(b1.nextOffset, 1, 87)
+      val b3 = createBatch(b2.nextOffset, 10, 912)
+      val b4 = createBatch(b3.nextOffset, 5, 200)
+
+      cache = cache.add(b1)
+      cache = cache.add(b2)
+      cache = cache.add(b3)
+
+      // Verify
+      val batch = cache.getAll(20,1100).get
+      batch.size shouldEqual b1.size + b2.size + b3.size
+      batch.sizeInBytes shouldEqual b1.sizeInBytes + b2.sizeInBytes + b3.sizeInBytes
+      batch.offset shouldEqual b1.offset
+      batch.nextOffset shouldEqual b3.nextOffset
+    }
```

Comment:
Add more tests.  I am going to use 500MB and 16MB max buffer size with 1MB sized messages for example, but scale properly to tests.

1. Make sure we respect the max buffer size.  That is to say if the cache is holding 500MB of batches, and there are 500 batches of 1MB in there, with max buffer size of 16MB, the cache should only return 16 batches worth of data.
   1. Sub variant - make sure this works properly at the beginning offset
   1. Sub variant - make sure this works properly if it's returning the last 16MB worth of messages and it's exactly flush with the end, i.e. we get exactly the last 16 messages
   1. Sub variant - make sure this works properly if it's returning the last 16MB of messages but the offset is 1 off, so we only get the last 15 messages
   1. Sub variant - make sure if the max batch size is 16MB, and the offset is 1 from the end, then we return only 1 batch back
   1. Sub variant - this should work properly in the middle of the 500mb, that is to say with an offset half way into the cache

Reply:
Done